### PR TITLE
Added marketUrl argument with clap

### PR DESCRIPTION
**Pull request by simzzz**
NOTE:  The code got formatted differently I'm not sure why.

### CODE REVIEW

**Comment by elpiel** [comment]
LGTM, One thing though is that the proxy part still uses the hardcoded value:
https://github.com/AdExNetwork/adex-supermarket/blob/800f99f4369056af0c6b4721ba4da9db50cb4fe4/src/lib.rs#L105

File: src/lib.rs [hunk 1]
```diff
@@ -97,10 +98,11 @@ async fn handle(
                 .map(ToOwned::to_owned)
                 .unwrap_or_else(|| PathAndQuery::from_static(""));
 
+            let split_uri: Vec<&str> = market_uri.split("://").collect();
```
Comment (elpiel) [inline src/lib.rs]: Looks good and could be improved, but we can make it much simpler.
We can concat the market_url and path_and_query. We can do that with `format!("{}{}", market_url, path_and_query)` and we can use this to parse it into a URI: https://docs.rs/http/0.1.17/http/uri/struct.Uri.html#examples
`.parse::<Uri>()?`

**Review by elpiel — CHANGES_REQUESTED** [review]

File: src/lib.rs [hunk 2]
```diff
@@ -113,7 +117,7 @@ async fn handle(
     }
 }
 
-async fn spawn_fetch_campaigns(market_uri: &str) -> Result<Cache, reqwest::Error> {
+async fn spawn_fetch_campaigns(market_uri: &String) -> Result<Cache, reqwest::Error> {
```
Comment (elpiel) [inline src/lib.rs]: Because of the way Rust works, you don't need `&String`, you can leave it as `&str` and pass a `....(&someStringVar)` which will automatically use the `&String` as a `&str`

File: src/main.rs [hunk 3]
```diff
@@ -17,7 +17,7 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
         )
         .get_matches();
 
-    let market_url = cli.value_of("marketUrl").expect("No market URL provided!");
+    let market_url = cli.value_of("marketUrl").expect("No market URL provided!").to_string();
```
Comment (elpiel) [inline src/main.rs:20]: Good job at using `expect()`, since we do `require` the `marketUrl` this should be there (this is why we use `expect()`) if anything goes wrong this will panic (crashing the program)

File: src/lib.rs [hunk 4]
```diff
@@ -47,21 +49,29 @@ impl From<reqwest::Error> for Error {
     }
 }
 
-pub async fn serve(addr: SocketAddr, market_url: &str) -> Result<(), Error> {
+impl From<http::uri::InvalidUri> for Error {
```
Comment (elpiel) [inline src/lib.rs:50]: You actually don't need this `InvalidUri`, the `http::Error` type implements `From<InvalidUri> for http::Error`, which means that using `?` will somatically use this From implementation and make the `InvalidUrl` into a `http::Error`.

**Hint:** We already have this error in our `enum Error` :raised_hands: so we can safely  remove the `Error::InvalidUri` from the enum and change this `impl From` to use `Error::Http(e.into())`.

**PS:** From & Into are very similar, you can read more about them in the docs https://doc.rust-lang.org/std/convert/trait.From.html
basically you can do `Error::from(http_error)` and it will use the `From<http::Error>`, but you can also just use `http_error.into()` if the compiler can see for example the return type of the function or in our case the `Error::Http` that it requires `http::Error` and will automatically do the correct conversion

**Review by elpiel — CHANGES_REQUESTED** [review]
You can see that From is impl in the docs here: https://docs.rs/http/0.2.1/http/uri/struct.Uri.html#implementations

File: src/lib.rs [hunk 5]
```diff
@@ -99,12 +108,7 @@ async fn handle(
                 .map(ToOwned::to_owned)
                 .unwrap_or_else(|| PathAndQuery::from_static(""));
 
-            let uri = Uri::builder()
-                // @TODO: Move to config or env. variable
-                .scheme("http")
-                .authority(Authority::from_static("localhost:8005"))
-                .path_and_query(path_and_query)
-                .build()?;
+            let uri = format!("{}{}", market_uri, path_and_query).parse::<Uri>()?;
```
Comment (elpiel) [inline src/lib.rs:109]: have you checked if `path_and_query` will add automatically `/`? maybe we can prevent errors from not passing ending `/` to the `market_uri` as we won't know if the value is ok until we actually make a request to the application.

quick solution will be to do `trim_end_matches('/')` and we will always expect the string will not have `/` (since we use it in multiple places) before we `to_string()` it in `main.rs`

E.g. we use it here: https://github.com/AdExNetwork/adex-supermarket/pull/3/files#diff-946d2bd7dec91649b58583b402d6d6c3R59
